### PR TITLE
Bug 721760 and Bug 774233: try to load vendor specific Sync settings for...

### PR DESCRIPTION
**Pull request by ncalexan**
... Moto Blur and HTC devices.

### CODE REVIEW

File: src/main/java/org/mozilla/gecko/sync/setup/SyncAccounts.java [hunk 1]
```diff
@@ -366,38 +366,68 @@ public static void setSyncAutomatically(Account account, boolean syncAutomatical
...
+   * @param context
+   *          current Android context.
+   * @return the <code>Intent</code> started.
+   */
+  public static Intent openSyncSettings(Context context) {
+    // Bug 721760 - opening Sync settings takes user to Battery & Data Manager
+    // on a variety of Motorola devices. This work around tries to load the
+    // correct Intent by hand. Oh, Android.
+    Intent intent;
+    if ((intent = openVendorSyncSettings(context, MOTO_BLUR_PACKAGE, MOTO_BLUR_SETTINGS_ACTIVITY)) != null) {
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/sync/setup/SyncAccounts.java]: Split out the assignment from the conditional for clarity, please.

File: src/main/java/org/mozilla/gecko/sync/setup/SyncAccounts.java [hunk 2]
```diff
@@ -366,38 +366,68 @@ public static void setSyncAutomatically(Account account, boolean syncAutomatical
...
+      return intent;
     }
 
     // Open default Sync settings activity.
-    final Intent intent = new Intent(Settings.ACTION_SYNC_SETTINGS);
-    intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
+    intent = new Intent(Settings.ACTION_SYNC_SETTINGS);
+    // Bug 774233: do not start activity as a new task (second run fails on some HTC devices).
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/sync/setup/SyncAccounts.java:430]: Tell me this: why did we open it in a new task in the first place? Presumably that fixed a different bug…

**Comment by rnewman** [comment]
Looks good, modulo one nit and one question.

File: src/main/java/org/mozilla/gecko/sync/setup/SyncAccounts.java (hunk 2 above)
Comment (ncalexan) [inline src/main/java/org/mozilla/gecko/sync/setup/SyncAccounts.java:430]: I really don't know; reading the Android docs at https://developer.android.com/guide/components/tasks-and-back-stack.html, I have no idea why we wanted a new task.  Kicking this to liuche.

**Comment by ncalexan** [comment]
liuche: can you answer the above question?

**Comment by liuche** [comment]
Originally, when writing the Fennec-back code for opening Fennec after setup (SetupSuccessActivity), I was seeing duplicate activities in the back-stack. I added NEW_TASK to remove duplication*. Didn't think it would cause problems, so I was probably careless in copying it to other places as well.

Went through the flow again, and it turns out it's not necessary to launch Sync Settings from SyncSetupActivity using an intent, so I removed it.

*from the docs for FLAG_ACTIVITY_NEW_TASK:
"When using this flag, if a task is already running for the activity you are now starting, then a new activity will not be started; instead, the current task will simply be brought to the front of the screen with the state it was last in."

**Comment by ncalexan** [comment]
nit: no bug number in commit comment.

nit: I'd like to add docstring to the two click handlers making clear that they are referenced from `android:onClick` in xml (which I hate, btw -- impossible to follow code flow with this style -- but I understand why we do this.)

non-nit: we have a bunch of stuff I don't trust (understand) to do with Android stack management.  In `manifests/SyncAndroidManifest_activities.xml.in`:
- one instance of android:taskAffinity="org.mozilla.gecko.sync.setup"
- two instances of android:clearTaskOnLaunch="true"
- two instances of android:noHistory="true"

Plus code specific hacks.  Do you understand all this and think that it is correct?  I find it bizarre that if we successfully set up, click on Fennec, then click back, we get the home screen.  Why are we setting noHistory at all?

**Comment by ncalexan** [comment]
After Vidyo with liuche, I'm happy with our Android stack management.  I'm going to get this onto m-b.

**Comment by ncalexan** [comment]
I want to keep this minimal, since we're landing on m-b, so I opened a new ticket for removing the new task flag more generally:

https://bugzilla.mozilla.org/show_bug.cgi?id=776659

and I pushed liuche's commit to a new tree:

https://github.com/mozilla-services/android-sync/tree/nalexander/bug-776659-remove-new-task-flag/

liuche, can you land this?

**Pull request merged**